From 2519c7d2fcdc8992c6091679f48d67dc657b0a40 Mon Sep 17 00:00:00 2001
From: Sunil Goutham <sgoutham@marvell.com>
Date: Mon, 27 Jan 2020 23:50:05 +0530
Subject: [PATCH 462/767] octeontx2-af: Fix rvu_reg_poll() to poll reg atleast
 twice

commit 62b92015e17be0209d4fcff476927b5223cd96ab from
git@git.assembla.com:cavium/WindRiver.linux.git

Currently if on the first check the operation is still not
finished, the poll goes to sleep for 2-5 usecs. But if for
some reason (due to secure interrupts etc) by the time the poll
wakes up the 10ms time is expired then we don't check if operation
is finished or not and return failure.

This patch fixes this to check HW operation after sleep so that
the status is checked atleast twice.

Change-Id: Ie05e21536b6a80738d4473617dad88e20f0b84eb
Signed-off-by: Sunil Goutham <sgoutham@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/22474
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/net/ethernet/marvell/octeontx2/af/rvu.c | 14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/af/rvu.c b/drivers/net/ethernet/marvell/octeontx2/af/rvu.c
index f9c2b0a23099..07fec147ddf1 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/rvu.c
+++ b/drivers/net/ethernet/marvell/octeontx2/af/rvu.c
@@ -90,13 +90,15 @@ int rvu_poll_reg(struct rvu *rvu, u64 block, u64 offset, u64 mask, bool zero)
 	u64 reg_val;
 
 	reg = rvu->afreg_base + ((block << 28) | offset);
-	while (time_before(jiffies, timeout)) {
-		reg_val = readq(reg);
-		if (zero && !(reg_val & mask))
-			return 0;
-		if (!zero && (reg_val & mask))
-			return 0;
+again:
+	reg_val = readq(reg);
+	if (zero && !(reg_val & mask))
+		return 0;
+	if (!zero && (reg_val & mask))
+		return 0;
+	if (time_before(jiffies, timeout)) {
 		usleep_range(1, 5);
+		goto again;
 	}
 	return -EBUSY;
 }
-- 
2.31.1

